docs(zone-sdk): user guide for bridging (#2898)
logos-co/nomos-node: zone-sdk/BRIDGING.md

# Bridging Assets Between the Bedrock and a Zone

> A user guide for zone developers using the Zone SDK to move tokens between the Logos blockchain (Bedrock) and a zone over a channel.

Reviewer: Bedrock is just an internal name, just use Logos Blockchain
```suggestion
A user guide for zone developers using the Zone SDK to move tokens between the Logos Blockchain and a zone over a channel.
```

## What "bridging" means here

> A **channel** on Bedrock is both the message log a zone publishes to *and* the bridge between Bedrock-side balances and zone-side balances. Each channel carries an on-chain balance ([`ChannelState.balance`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#22b261aa09df8289a3f281de4aa8fdca)); deposits credit it, withdrawals debit it. The zone is free to define how that balance maps to its own internal accounts — the SDK only surfaces the on-chain events.

Reviewer: ```suggestion
A **channel** on Logos Blockchain is both the message log a zone publishes to *and* the bridge between Logos Blockchain-side balances and Zone-side balances. Each channel carries an on-chain balance ([`ChannelState.balance`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#22b261aa09df8289a3f281de4aa8fdca)); deposits credit it, withdrawals debit it. The zone is free to define how that balance maps to its own internal accounts — the SDK only surfaces the on-chain events.
```

Two directions:

> - **Deposit** (Bedrock -> zone). A user spends notes on Bedrock into a channel via [`ChannelDeposit`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#80b261aa09df8353814a81efe0fbd8ed). The channel balance grows. The zone sequencer observes the finalized deposit and credits the user inside the zone according to `ChannelDeposit.metadata`.

Reviewer: ```suggestion
- **Deposit** (Blockchain -> Zone). A user deposits notes into a channel via [`ChannelDeposit`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#80b261aa09df8353814a81efe0fbd8ed). The channel balance grows. The zone sequencer observes the finalized deposit and credits the user inside the zone according to `ChannelDeposit.metadata`.
```

> - **Withdraw** (zone -> Bedrock). The zone sequencer submits [`ChannelWithdraw`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#cd7261aa09df83dd98b3017dafc37e87), signed by [`ChannelState.withdraw_threshold`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#22b261aa09df8289a3f281de4aa8fdca) accredited keys. The channel balance shrinks and fresh notes appear on Bedrock for the recipients named in `ChannelWithdraw.outputs`.

Reviewer: ```suggestion
- **Withdraw** (Zone -> Blockchain). The zone sequencer submits [`ChannelWithdraw`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#cd7261aa09df83dd98b3017dafc37e87), signed by [`ChannelState.withdraw_threshold`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#22b261aa09df8289a3f281de4aa8fdca) accredited keys. The channel balance shrinks and fresh notes appear on the Blockchain for the recipients named in `ChannelWithdraw.outputs`.
```

The Zone SDK is the client library for *zone-side* code: sequencers issuing withdraws and observing deposits, indexers replaying the message log.

## Creating a channel

> Channels are not deployed by a separate transaction; Bedrock creates them just-in-time on the first operation that references a previously unseen [`ChannelId`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#22b261aa09df8289a3f281de4aa8fdca). A `ChannelId` is a 32-byte identifier chosen by the creator. Whoever signs that first operation becomes the sole accredited key, and the channel starts with `configuration_threshold = 1`, `withdraw_threshold = 1`, and `balance = 0`.

Reviewer: ```suggestion
Channels are not deployed by a separate transaction; Channels are created just-in-time on the first operation that references a previously unseen [`ChannelId`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#22b261aa09df8289a3f281de4aa8fdca). A `ChannelId` is a 32-byte identifier chosen by the creator. Whoever signs that first operation becomes the sole accredited key, and the channel starts with `configuration_threshold = 1`, `withdraw_threshold = 1`, and `balance = 0`.
```

From the Zone SDK, the steps are:

1. Pick a `ChannelId` and a sequencer `Ed25519Key`.

> 2. Initialize a `ZoneSequencer`, drive it (see the SDK overview for the drive-loop pattern and the `Event::Ready` readiness contract), and publish the first inscription (e.g., a zone genesis block) via `sequencer.handle().publish(..)`. Bedrock creates the channel automatically, naming this sequencer as the sole accredited key.

Reviewer: ```suggestion
2. Initialize a `ZoneSequencer`, drive it (see the SDK overview for the drive-loop pattern and the `Event::Ready` readiness contract), and publish the first inscription (e.g., a zone genesis block) via `sequencer.handle().publish(..)`. On the Blockchain, Channels are crated automatically, naming this sequencer as the sole accredited key.
```

`BlocksProcessed` fires per ingested block (live or backfill) and only carries finalized items at or below LIB, so deposits surfaced here cannot be re-orged off the chain.

```rust
use lb_zone_sdk::sequencer::{Event, FinalizedOp};

if let Event::BlocksProcessed { finalized, .. } = event {
    for tx in finalized {
        for op in tx.ops {
            if let FinalizedOp::Deposit(deposit) = op {
                println!(
                    "Deposit of {} with metadata {:?}",
                    deposit.amount, deposit.metadata,
                );
            }
        }
    }
}
```

## Withdrawals: zone -> Bedrock

A withdraw is initiated *inside the zone* and lands on Bedrock as a signed [`ChannelWithdraw`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#5de261aa09df8321b05401f2e8dea08b) operation.

The `ChannelWithdraw` specifies the `channel`, the `outputs` (new notes to mint on Bedrock), and the `withdraw_nonce` that must match `ChannelState.withdrawal_nonce` for replay protection.

> The `ChannelWithdrawOpProof` carries `ChannelState.withdraw_threshold` signatures from distinct `ChannelState.accredited_keys`.

Reviewer: ```suggestion
## Withdrawals: Zone -> Blockchain

A withdraw is initiated *inside the zone* and lands on-chain as a signed [`ChannelWithdraw`](https://app.notion.com/p/nomos-tech/1-5-0-Mantle-33d261aa09df8051b0d0cd4d5ddade85?source=copy_link#5de261aa09df8321b05401f2e8dea08b) operation.

The `ChannelWithdraw` specifies the `channel`, the `outputs` (new notes to mint on-chain), and the `withdraw_nonce` that must match `ChannelState.withdrawal_nonce` for replay protection.

The `ChannelWithdrawOpProof` carries `ChannelState.withdraw_threshold` signatures from distinct `ChannelState.accredited_keys`.
```

### Single-sequencer zones

Currently, the Zone SDK supports the bundled withdrawal API only for single-sequencer zones (`ChannelState.withdraw_threshold == 1`). You describe *what* to withdraw via a `WithdrawArg` — just the recipient `Outputs`. The SDK fills in the `channel_id` and reads the current `withdraw_nonce` and this sequencer's accredited-key index from cached channel state.

```rust
use lb_core::mantle::{Note, ledger::Outputs};
use lb_zone_sdk::sequencer::WithdrawArg;

let withdraw = WithdrawArg {
    outputs: Outputs::new([Note::new(50, recipient_pk)]),
};

// Inside the drive task: submit the inscription bundled with the withdraw.
let (result, checkpoint) = sequencer.handle().publish_atomic_withdraw(
    inscription_payload,   // the zone block this withdraw goes with
    vec![withdraw],
)?;
```

Because the inscription and the withdraw share one transaction, they adopt/orphan/finalize as a unit — the zone block recording the withdraw and the on-chain debit cannot drift apart.

#### Observing the result

> `publish_atomic_withdraw` returns the `PublishResult` synchronously. For a single-sig bundle, `PublishResult.tx` is a `PendingTx::AtomicWithdraw(AtomicWithdrawInfo)` carrying the inscription and the bundled withdraw ops. Persist it immediately as your outbox.

Reviewer: is this still true? I thought this API was sync now.

Reviewer 2: Maybe the wording is not the best? It does say it returns PublishResult synchronously -> there's no separate Published event anymore, so the bundle info comes back on the return value.

Reviewer 2: Maybe it should just be removed from this section all together, as publish call return result right away

Reviewer: ah no, sorry was reading it too quickly, your right, it's fine

let signatures: Vec<IndexedSignature> = collect_signatures_from_committee(
    &tx,
    IndexedSignature::new(own_key_index, own_sig.clone()),
).await?;

// 4. Assemble the threshold proof and submit.
let withdraw_proof = ChannelMultiSigProof::new(signatures)?;
let signed_tx = SignedMantleTx::new(
    tx,
    vec![
        OpProof::ChannelMultiSigProof(withdraw_proof),
        OpProof::Ed25519Sig(own_sig),
    ],
)?;
let (result, checkpoint) = sequencer
    .handle()
    .submit_signed_tx(signed_tx, msg_id)?;
```

#### Observing the result

`submit_signed_tx` returns the `PublishResult` synchronously. Unlike the single-sig flow, the SDK treats the caller-built tx as opaque, so `PublishResult.tx` is `PendingTx::Inscription(InscriptionInfo)` regardless of the underlying ops — the bundle structure is not reconstructed in the result. Persist it as your outbox using the returned `tx_hash` to identify the bundle.

The finalization pattern is the same as in the single-sig case: `Event::BlocksProcessed.finalized` carries `FinalizedOp::Inscription` and `FinalizedOp::Withdraw` for the bundle in one `tx.ops`. Match by `tx_hash` and iterate both ops as shown above.

### Reorgs and republish

> If a withdraw submitted via `publish_atomic_withdraw` has its parent inscription orphaned by a chain reorg, the SDK reports it via the `channel_update` field of `Event::BlocksProcessed`, with the abandoned tx in `channel_update.orphaned`. The original signed transaction is no longer valid. The consumer decides whether to republish — re-call `publish_atomic_withdraw` with the same inscription payload and `WithdrawArg`s reconstructed from the bundle; the SDK refills the inscription parent and the `withdraw_nonce` from current Bedrock state.

Reviewer: ```suggestion
If a withdraw submitted via `publish_atomic_withdraw` has its parent inscription orphaned by a chain reorg, the SDK reports it via the `channel_update` field of `Event::BlocksProcessed`, with the abandoned tx in `channel_update.orphaned`. The original signed transaction is no longer valid. The consumer decides whether to republish — re-call `publish_atomic_withdraw` with the same inscription payload and `WithdrawArg`s reconstructed from the bundle; the SDK refills the inscription parent and the `withdraw_nonce` from current on-chain state.
```
